src/cls/rbd/cls_rbd_types.h

```diff
@@ -495,6 +507,69 @@ struct TrashSnapshotNamespace {
   }
 };
 
+struct MirrorPrimarySnapshotNamespace {
+  static const SnapshotNamespaceType SNAPSHOT_NAMESPACE_TYPE =
+    SNAPSHOT_NAMESPACE_TYPE_MIRROR_PRIMARY;
+
+  bool demoted = false;
+  std::set<std::string> mirror_peer_uuids;
+
+  MirrorPrimarySnapshotNamespace() {
+  }
+  MirrorPrimarySnapshotNamespace(bool demoted,
+                                 const std::set<std::string> &mirror_peer_uuids)
+    : demoted(demoted), mirror_peer_uuids(mirror_peer_uuids) {
+  }
+
+  void encode(bufferlist& bl) const;
+  void decode(bufferlist::const_iterator& it);
+
+  void dump(Formatter *f) const;
+
+  inline bool operator==(const MirrorPrimarySnapshotNamespace& mpsn) const {
```

Comment:
The idea was to compare only "immutable" fields, I thought it could be useful if we want to see the snapshot is the same, e.g. after unlinking a peer. Anyway, right now it is never used (may be in tests only) so I will make it less confusing.